<launch>
  <!-- Enable/disable and connect toggles per arm -->
  <arg name="enable_jaka1" default="true" />
  <arg name="enable_jaka2" default="false" />
  <arg name="enable_jaka3" default="false" />
  <arg name="enable_jaka4" default="false" />

  <arg name="connect_jaka1" default="true" />
  <arg name="connect_jaka2" default="true" />
  <arg name="connect_jaka3" default="true" />
  <arg name="connect_jaka4" default="true" />

  <!-- Per-arm calibration and input topics -->
  <arg name="calib_yaml_jaka1" default="$(find jaka_close_contro)/config/world_robot_extrinsic_offline_jaka1.yaml" />
  <arg name="calib_yaml_jaka2" default="$(find jaka_close_contro)/config/world_robot_extrinsic_offline_jaka2.yaml" />
  <arg name="calib_yaml_jaka3" default="$(find jaka_close_contro)/config/world_robot_extrinsic_offline_jaka3.yaml" />
  <arg name="calib_yaml_jaka4" default="$(find jaka_close_contro)/config/world_robot_extrinsic_offline_jaka4.yaml" />

  <arg name="cube_pose_topic_jaka1" default="/vision/jaka1/cube_center_world" />
  <arg name="cube_pose_topic_jaka2" default="/vision/jaka2/cube_center_world" />
  <arg name="cube_pose_topic_jaka3" default="/vision/jaka3/cube_center_world" />
  <arg name="cube_pose_topic_jaka4" default="/vision/jaka4/cube_center_world" />

  <!-- Vision pipeline toggle -->
  <arg name="start_vision" default="true" />

  <!-- Vision/common parameters -->
  <arg name="robot_ids" default="[1,2,3,4]" />
  <arg name="faces_yaml" default="$(find jaka_close_contro)/config/cube_faces_4robots.yaml" />
  <arg name="svo_file" default="" />
  <arg name="world_frame" default="world" />
  <arg name="camera_frame" default="zed2i_left_camera_optical_frame" />
  <arg name="urdf_file" default="$(find jaka_close_contro)/urdf/jaka_zu3.urdf" />
  <arg name="strict_cube_frame" default="true" />

  <!-- Robot IPs -->
  <arg name="jaka1_ip" default="192.168.1.100" />
  <arg name="jaka2_ip" default="192.168.1.101" />
  <arg name="jaka3_ip" default="192.168.1.102" />
  <arg name="jaka4_ip" default="192.168.1.103" />

  <!-- Control parameters -->
  <arg name="control_rate_hz" default="20.0" />
  <arg name="kp_pos" default="0.6" />
  <arg name="kp_ang" default="0.8" />
  <arg name="v_max" default="0.02" />
  <arg name="w_max_deg" default="10.0" />
  <arg name="eps_pos_m" default="0.0015" />
  <arg name="eps_ang_deg" default="0.5" />
  <arg name="vision_timeout_s" default="0.2" />
  <arg name="servo_timeout_s" default="10.0" />

  <!-- 0. Robot drivers (namespaced) -->
  <group ns="jaka1" if="$(arg enable_jaka1)">
    <param name="ip" value="$(arg jaka1_ip)" />
    <param name="robot_description" textfile="$(arg urdf_file)" />

    <node pkg="jaka_driver"
          type="jaka_driver"
          name="jaka_driver"
          output="screen">
      <param name="ip" value="$(arg jaka1_ip)" />
      <param name="robot_ip" value="$(arg jaka1_ip)" />
    </node>

    <node pkg="jaka_close_contro"
          type="jaka_state_adapter_node"
          name="jaka_state_adapter"
          output="screen" />

    <node pkg="robot_state_publisher"
          type="robot_state_publisher"
          name="robot_state_publisher">
      <param name="publish_frequency" value="100.0" />
    </node>
  </group>

  <group ns="jaka2" if="$(arg enable_jaka2)">
    <param name="ip" value="$(arg jaka2_ip)" />
    <param name="robot_description" textfile="$(arg urdf_file)" />

    <node pkg="jaka_driver"
          type="jaka_driver"
          name="jaka_driver"
          output="screen">
      <param name="ip" value="$(arg jaka2_ip)" />
      <param name="robot_ip" value="$(arg jaka2_ip)" />
    </node>

    <node pkg="jaka_close_contro"
          type="jaka_state_adapter_node"
          name="jaka_state_adapter"
          output="screen" />

    <node pkg="robot_state_publisher"
          type="robot_state_publisher"
          name="robot_state_publisher">
      <param name="publish_frequency" value="100.0" />
    </node>
  </group>

  <group ns="jaka3" if="$(arg enable_jaka3)">
    <param name="ip" value="$(arg jaka3_ip)" />
    <param name="robot_description" textfile="$(arg urdf_file)" />

    <node pkg="jaka_driver"
          type="jaka_driver"
          name="jaka_driver"
          output="screen">
      <param name="ip" value="$(arg jaka3_ip)" />
      <param name="robot_ip" value="$(arg jaka3_ip)" />
    </node>

    <node pkg="jaka_close_contro"
          type="jaka_state_adapter_node"
          name="jaka_state_adapter"
          output="screen" />

    <node pkg="robot_state_publisher"
          type="robot_state_publisher"
          name="robot_state_publisher">
      <param name="publish_frequency" value="100.0" />
    </node>
  </group>

  <group ns="jaka4" if="$(arg enable_jaka4)">
    <param name="ip" value="$(arg jaka4_ip)" />
    <param name="robot_description" textfile="$(arg urdf_file)" />

    <node pkg="jaka_driver"
          type="jaka_driver"
          name="jaka_driver"
          output="screen">
      <param name="ip" value="$(arg jaka4_ip)" />
      <param name="robot_ip" value="$(arg jaka4_ip)" />
    </node>

    <node pkg="jaka_close_contro"
          type="jaka_state_adapter_node"
          name="jaka_state_adapter"
          output="screen" />

    <node pkg="robot_state_publisher"
          type="robot_state_publisher"
          name="robot_state_publisher">
      <param name="publish_frequency" value="100.0" />
    </node>
  </group>

  <!-- 1. Vision pipeline -->
  <group if="$(arg start_vision)">
    <!-- ZED2i camera -->
    <include file="$(find zed_wrapper)/launch/zed2i.launch">
      <arg name="svo_file" value="$(arg svo_file)" />
    </include>

    <!-- ArUco cube detector -->
    <node pkg="jaka_close_contro"
          type="cube_aruco_detector_node"
          name="cube_aruco_detector"
          output="screen">
      <param name="image_topic"       value="/zed2i/zed_node/left/image_rect_color"/>
      <param name="camera_info_topic" value="/zed2i/zed_node/left/camera_info"/>
      <param name="fiducial_topic"    value="/fiducial_transforms"/>
      <param name="default_marker_size" value="0.08"/>
      <param name="dictionary" value="DICT_4X4_50"/>
      <param name="tool_dictionary" value="DICT_6X6_1000"/>
      <param name="tool_marker_length_m" value="0.0425"/>
      <param name="tool_marker_separation_m" value="0.005"/>
      <param name="min_gridboard_markers" value="2"/>
      <param name="world_board_enable" value="true"/>
      <param name="world_board_output_id" value="500"/>
      <param name="world_board_marker_length_m" value="0.0525"/>
      <param name="world_board_marker_separation_m" value="0.005"/>
      <param name="world_board_min_markers" value="2"/>
      <param name="enable_single_world_markers" value="false"/>
      <param name="robot_stride" value="100" />
      <param name="id_step" value="10" />
      <param name="face_id_base" value="10" />
      <rosparam subst_value="true" param="robot_ids">$(arg robot_ids)</rosparam>
      <rosparam param="marker_sizes">
        "0": 0.08  <!-- World Tag (80mm) -->
      </rosparam>
      <rosparam param="world_board_marker_ids">[500, 501, 502, 503]</rosparam>
    </node>

    <!-- Fiducial relay -->
    <node pkg="jaka_close_contro"
          type="fiducial_relay_node"
          name="fiducial_relay"
          output="screen">
      <rosparam param="world_ids">[500, 501, 502, 503]</rosparam>
      <rosparam subst_value="true" param="robot_ids">$(arg robot_ids)</rosparam>
      <param name="input_topic" value="/fiducial_transforms"/>
      <param name="output_world_topic" value="/world_fiducials"/>
      <param name="output_tool_topic"  value="/tool_fiducials"/>
    </node>

    <!-- World tag estimator -->
    <node pkg="jaka_close_contro"
          type="world_tag_node"
          name="world_tag_node"
          output="screen">
      <rosparam param="world_ids">[500, 501, 502, 503]</rosparam>
      <param name="fiducial_topic"    value="/world_fiducials"/>
      <param name="world_frame"       value="$(arg world_frame)"/>
      <param name="camera_frame"      value="$(arg camera_frame)"/>
      <param name="alpha_pos"         value="0.3"/>
      <param name="alpha_rot"         value="0.3"/>
    </node>

    <!-- Multi-face fusion per robot -->
    <group if="$(arg enable_jaka1)">
      <node pkg="jaka_close_contro"
            type="cube_multi_face_fusion_node"
            name="cube_multi_face_fusion_jaka1"
            output="screen">
        <param name="faces_yaml"           value="$(arg faces_yaml)"/>
        <param name="fiducial_topic"       value="/tool_fiducials"/>
        <param name="cube_frame"           value="cube_center_jaka1"/>
        <param name="camera_frame_default" value="$(arg camera_frame)"/>
        <param name="timeout_sec"          value="0.5"/>
        <param name="robot_id" value="1" />
        <param name="robot_name" value="jaka1" />
        <param name="face_id_base" value="10" />
        <param name="robot_stride" value="100" />
        <param name="output_topic" value="/vision/jaka1/cube_center" />
        <param name="stats_topic" value="/vision/jaka1/cube_fusion_stats" />
        <param name="filter_robot_faces" value="true" />
        <param name="apply_robot_offset" value="false" />
        <param name="broadcast_tf" value="true" />
      </node>
    </group>

    <group if="$(arg enable_jaka2)">
      <node pkg="jaka_close_contro"
            type="cube_multi_face_fusion_node"
            name="cube_multi_face_fusion_jaka2"
            output="screen">
        <param name="faces_yaml"           value="$(arg faces_yaml)"/>
        <param name="fiducial_topic"       value="/tool_fiducials"/>
        <param name="cube_frame"           value="cube_center_jaka2"/>
        <param name="camera_frame_default" value="$(arg camera_frame)"/>
        <param name="timeout_sec"          value="0.5"/>
        <param name="robot_id" value="2" />
        <param name="robot_name" value="jaka2" />
        <param name="face_id_base" value="10" />
        <param name="robot_stride" value="100" />
        <param name="output_topic" value="/vision/jaka2/cube_center" />
        <param name="stats_topic" value="/vision/jaka2/cube_fusion_stats" />
        <param name="filter_robot_faces" value="true" />
        <param name="apply_robot_offset" value="false" />
        <param name="broadcast_tf" value="true" />
      </node>
    </group>

    <group if="$(arg enable_jaka3)">
      <node pkg="jaka_close_contro"
            type="cube_multi_face_fusion_node"
            name="cube_multi_face_fusion_jaka3"
            output="screen">
        <param name="faces_yaml"           value="$(arg faces_yaml)"/>
        <param name="fiducial_topic"       value="/tool_fiducials"/>
        <param name="cube_frame"           value="cube_center_jaka3"/>
        <param name="camera_frame_default" value="$(arg camera_frame)"/>
        <param name="timeout_sec"          value="0.5"/>
        <param name="robot_id" value="3" />
        <param name="robot_name" value="jaka3" />
        <param name="face_id_base" value="10" />
        <param name="robot_stride" value="100" />
        <param name="output_topic" value="/vision/jaka3/cube_center" />
        <param name="stats_topic" value="/vision/jaka3/cube_fusion_stats" />
        <param name="filter_robot_faces" value="true" />
        <param name="apply_robot_offset" value="false" />
        <param name="broadcast_tf" value="true" />
      </node>
    </group>

    <group if="$(arg enable_jaka4)">
      <node pkg="jaka_close_contro"
            type="cube_multi_face_fusion_node"
            name="cube_multi_face_fusion_jaka4"
            output="screen">
        <param name="faces_yaml"           value="$(arg faces_yaml)"/>
        <param name="fiducial_topic"       value="/tool_fiducials"/>
        <param name="cube_frame"           value="cube_center_jaka4"/>
        <param name="camera_frame_default" value="$(arg camera_frame)"/>
        <param name="timeout_sec"          value="0.5"/>
        <param name="robot_id" value="4" />
        <param name="robot_name" value="jaka4" />
        <param name="face_id_base" value="10" />
        <param name="robot_stride" value="100" />
        <param name="output_topic" value="/vision/jaka4/cube_center" />
        <param name="stats_topic" value="/vision/jaka4/cube_fusion_stats" />
        <param name="filter_robot_faces" value="true" />
        <param name="apply_robot_offset" value="false" />
        <param name="broadcast_tf" value="true" />
      </node>
    </group>
  </group>

  <!-- 2. TF -> Pose bridge for each robot -->
  <group if="$(arg enable_jaka1)">
    <node pkg="jaka_close_contro"
          type="tf_to_pose_stamped_node.py"
          name="tf_to_pose_stamped_jaka1"
          output="screen">
      <param name="parent_frame" value="$(arg world_frame)" />
      <param name="child_frame" value="cube_center_jaka1" />
      <param name="output_topic" value="/vision/jaka1/cube_center_world" />
      <param name="rate_hz" value="30.0" />
    </node>
  </group>

  <group if="$(arg enable_jaka2)">
    <node pkg="jaka_close_contro"
          type="tf_to_pose_stamped_node.py"
          name="tf_to_pose_stamped_jaka2"
          output="screen">
      <param name="parent_frame" value="$(arg world_frame)" />
      <param name="child_frame" value="cube_center_jaka2" />
      <param name="output_topic" value="/vision/jaka2/cube_center_world" />
      <param name="rate_hz" value="30.0" />
    </node>
  </group>

  <group if="$(arg enable_jaka3)">
    <node pkg="jaka_close_contro"
          type="tf_to_pose_stamped_node.py"
          name="tf_to_pose_stamped_jaka3"
          output="screen">
      <param name="parent_frame" value="$(arg world_frame)" />
      <param name="child_frame" value="cube_center_jaka3" />
      <param name="output_topic" value="/vision/jaka3/cube_center_world" />
      <param name="rate_hz" value="30.0" />
    </node>
  </group>

  <group if="$(arg enable_jaka4)">
    <node pkg="jaka_close_contro"
          type="tf_to_pose_stamped_node.py"
          name="tf_to_pose_stamped_jaka4"
          output="screen">
      <param name="parent_frame" value="$(arg world_frame)" />
      <param name="child_frame" value="cube_center_jaka4" />
      <param name="output_topic" value="/vision/jaka4/cube_center_world" />
      <param name="rate_hz" value="30.0" />
    </node>
  </group>

  <!-- 3. Pose servo nodes -->
  <group ns="jaka1" if="$(arg enable_jaka1)">
    <node pkg="jaka_close_contro" type="pose_servo_world_node" name="pose_servo_world" output="screen">
      <param name="robot_name" value="jaka1" />
      <param name="robot_id" value="1" />
      <param name="enable" value="$(arg enable_jaka1)" />
      <param name="connect_robot" value="$(arg connect_jaka1)" />
      <param name="calib_yaml" value="$(arg calib_yaml_jaka1)" />
      <param name="cube_pose_topic" value="$(arg cube_pose_topic_jaka1)" />
      <param name="target_topic" value="target_tool_world" />
      <param name="control_rate_hz" value="$(arg control_rate_hz)" />
      <param name="kp_pos" value="$(arg kp_pos)" />
      <param name="kp_ang" value="$(arg kp_ang)" />
      <param name="v_max" value="$(arg v_max)" />
      <param name="w_max_deg" value="$(arg w_max_deg)" />
      <param name="eps_pos_m" value="$(arg eps_pos_m)" />
      <param name="eps_ang_deg" value="$(arg eps_ang_deg)" />
      <param name="vision_timeout_s" value="$(arg vision_timeout_s)" />
      <param name="servo_timeout_s" value="$(arg servo_timeout_s)" />
      <param name="world_frame" value="$(arg world_frame)" />
      <param name="strict_cube_frame" value="$(arg strict_cube_frame)" />
    </node>
  </group>

  <group ns="jaka2" if="$(arg enable_jaka2)">
    <node pkg="jaka_close_contro" type="pose_servo_world_node" name="pose_servo_world" output="screen">
      <param name="robot_name" value="jaka2" />
      <param name="robot_id" value="2" />
      <param name="enable" value="$(arg enable_jaka2)" />
      <param name="connect_robot" value="$(arg connect_jaka2)" />
      <param name="calib_yaml" value="$(arg calib_yaml_jaka2)" />
      <param name="cube_pose_topic" value="$(arg cube_pose_topic_jaka2)" />
      <param name="target_topic" value="target_tool_world" />
      <param name="control_rate_hz" value="$(arg control_rate_hz)" />
      <param name="kp_pos" value="$(arg kp_pos)" />
      <param name="kp_ang" value="$(arg kp_ang)" />
      <param name="v_max" value="$(arg v_max)" />
      <param name="w_max_deg" value="$(arg w_max_deg)" />
      <param name="eps_pos_m" value="$(arg eps_pos_m)" />
      <param name="eps_ang_deg" value="$(arg eps_ang_deg)" />
      <param name="vision_timeout_s" value="$(arg vision_timeout_s)" />
      <param name="servo_timeout_s" value="$(arg servo_timeout_s)" />
      <param name="world_frame" value="$(arg world_frame)" />
      <param name="strict_cube_frame" value="$(arg strict_cube_frame)" />
    </node>
  </group>

  <group ns="jaka3" if="$(arg enable_jaka3)">
    <node pkg="jaka_close_contro" type="pose_servo_world_node" name="pose_servo_world" output="screen">
      <param name="robot_name" value="jaka3" />
      <param name="robot_id" value="3" />
      <param name="enable" value="$(arg enable_jaka3)" />
      <param name="connect_robot" value="$(arg connect_jaka3)" />
      <param name="calib_yaml" value="$(arg calib_yaml_jaka3)" />
      <param name="cube_pose_topic" value="$(arg cube_pose_topic_jaka3)" />
      <param name="target_topic" value="target_tool_world" />
      <param name="control_rate_hz" value="$(arg control_rate_hz)" />
      <param name="kp_pos" value="$(arg kp_pos)" />
      <param name="kp_ang" value="$(arg kp_ang)" />
      <param name="v_max" value="$(arg v_max)" />
      <param name="w_max_deg" value="$(arg w_max_deg)" />
      <param name="eps_pos_m" value="$(arg eps_pos_m)" />
      <param name="eps_ang_deg" value="$(arg eps_ang_deg)" />
      <param name="vision_timeout_s" value="$(arg vision_timeout_s)" />
      <param name="servo_timeout_s" value="$(arg servo_timeout_s)" />
      <param name="world_frame" value="$(arg world_frame)" />
      <param name="strict_cube_frame" value="$(arg strict_cube_frame)" />
    </node>
  </group>

  <group ns="jaka4" if="$(arg enable_jaka4)">
    <node pkg="jaka_close_contro" type="pose_servo_world_node" name="pose_servo_world" output="screen">
      <param name="robot_name" value="jaka4" />
      <param name="robot_id" value="4" />
      <param name="enable" value="$(arg enable_jaka4)" />
      <param name="connect_robot" value="$(arg connect_jaka4)" />
      <param name="calib_yaml" value="$(arg calib_yaml_jaka4)" />
      <param name="cube_pose_topic" value="$(arg cube_pose_topic_jaka4)" />
      <param name="target_topic" value="target_tool_world" />
      <param name="control_rate_hz" value="$(arg control_rate_hz)" />
      <param name="kp_pos" value="$(arg kp_pos)" />
      <param name="kp_ang" value="$(arg kp_ang)" />
      <param name="v_max" value="$(arg v_max)" />
      <param name="w_max_deg" value="$(arg w_max_deg)" />
      <param name="eps_pos_m" value="$(arg eps_pos_m)" />
      <param name="eps_ang_deg" value="$(arg eps_ang_deg)" />
      <param name="vision_timeout_s" value="$(arg vision_timeout_s)" />
      <param name="servo_timeout_s" value="$(arg servo_timeout_s)" />
      <param name="world_frame" value="$(arg world_frame)" />
      <param name="strict_cube_frame" value="$(arg strict_cube_frame)" />
    </node>
  </group>
</launch>
